bin/reflective_loader/src/loader.rs

```diff
@@ -0,0 +1,844 @@
+    if library_handle == 0 {
+        let error_code = unsafe { err_ptr() };
+        if error_code != 0 {
+            panic!("Failed to load library. Last error returned: {}", error_code);
+        }
+    }
+    library_handle
+}
+
+#[derive(Debug, Copy, Clone)]
+struct BaseRelocationEntry {
+    offset: u16,
+    reloc_type: u16,
+}
+
+impl BaseRelocationEntry {
+    fn new(c_bytes: u16) -> Self {
+        let reloc_type_bit_mask: u16 = 0b1111_0000_0000_0000;
+        let reloc_type = (c_bytes & reloc_type_bit_mask) >> 12;
+        
+        let offset_bit_mask: u16 = 0b0000_1111_1111_1111;
+        let offset = c_bytes & offset_bit_mask;
+        
+        Self {
+            offset,
+            reloc_type,
+        }
+    }
+
+    fn c_size() -> usize {
+        core::mem::size_of::<u16>()
+    }
+}
+
+#[cfg(target_arch = "x86_64")]
+struct PeFileHeaders64 {
+    dos_headers: IMAGE_DOS_HEADER,
+    nt_headers: IMAGE_NT_HEADERS64,
+    section_headers: [IMAGE_SECTION_HEADER; MAX_PE_SECTIONS],
+}
+
+// Pares the PE file from a series of bytes
+#[cfg(target_arch = "x86_64")]
+impl PeFileHeaders64 {
+    fn new(dll_bytes_ptr: *mut c_void) -> Self{
+        // DOS Headers
+        let dos_headers_base_ptr = dll_bytes_ptr as usize;
+        let dos_headers = unsafe { *((dos_headers_base_ptr) as *mut IMAGE_DOS_HEADER) };
+        if dos_headers.e_magic != PE_MAGIC {
+            panic!("PE Magic header mismatch. Expected 0x5A4D == MZ == 21117. File does not appear to be a PE executable.");
+        }
+
+        // NT Headers
+        let nt_headers_base_ptr = dos_headers_base_ptr + dos_headers.e_lfanew as usize;
+        let nt_headers = unsafe { *((nt_headers_base_ptr) as *mut IMAGE_NT_HEADERS64) };
+        if nt_headers.Signature != NT_SIGNATURE {
+            panic!("NT Signature mismatch. Expected 0x4550 == PE == 17744. File does not appear to be a PE executable.");
+        }
+
+        // Section Headers - hopefully there isn't more than MAX_PE_SECTIONS sections.
+        let null_section = IMAGE_SECTION_HEADER{ 
+            Name: [0; 8], 
+            Misc: IMAGE_SECTION_HEADER_0 { 
+                PhysicalAddress: 0, 
+            },
+            VirtualAddress: 0, 
+            SizeOfRawData: 0, 
+            PointerToRawData: 0, 
+            PointerToRelocations: 0, 
+            PointerToLinenumbers: 0, 
+            NumberOfRelocations: 0, 
+            NumberOfLinenumbers: 0, 
+            Characteristics: 0
+        };
+        let mut section_headers: [IMAGE_SECTION_HEADER; MAX_PE_SECTIONS] = [null_section; MAX_PE_SECTIONS];
+        let optional_headers_start_ptr = unsafe{&(*(nt_headers_base_ptr as *mut IMAGE_NT_HEADERS64)).OptionalHeader as *const _ as usize};
+        let section_headers_start_ptr = optional_headers_start_ptr + nt_headers.FileHeader.SizeOfOptionalHeader as usize;
+        let mut cur_section_ptr = section_headers_start_ptr as *mut IMAGE_SECTION_HEADER;
```

Comment:
I really recommend making a `core::slice` using `from_raw_parts` then you can just `copy_from_slice` into your array.

Reply:
So I see this example in around the same place in memN0ps code but not sure how that would play into my code.
https://github.com/memN0ps/rdi-rs/blob/5d7dd15d78030d0c56aaed4e9f26a7cae4c9cf88/reflective_loader/src/loader.rs#L140C13-L140C24


Could you provide an example of what you're thinking?

Reply:
```rust
let section_headers = unsafe { core::slice::from_raw_parts(section_headers_start_ptr as *const IMAGE_SECTION_HEADER, nt_headers.FileHeader.NumberOfSections as usize) };
for sec in section_headers.iter() {
    // do things with each section
}
```

Reply:
really, you should be referring to the headers by reference and not cloning them.